### PR TITLE
fix: catch PartsError and ProviderError from dispatch run

### LINKED ISSUE #55: Errors/tracebacks are not being correctly propagate to the cli/log
### Bug Description

Errors generated by the craft-libraries are being [silently discarded](https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L255) when sending the error to the emitter. The result is a single "<appname> internal error: <message>" in the cli and log

### To Reproduce

Run any project that fails to build in craft-parts (or through a programming error).

### part yaml

_No response_

### Relevant log output

```shell
none
```

### PR DESCRIPTION
**Pull request by sergiusens**
Capture these errors so they are not treated as internal errors with no details about the error.

- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----
Fixes #55

### CODE REVIEW

File: craft_application/application.py [hunk 1]
```diff
@@ -294,6 +295,20 @@ def run(self) -> int:
...
+                )
+            )
+            return_code = 256
+        except craft_providers.ProviderError as err:
+            self._emit_error(
+                craft_cli.CraftError(
+                    err.brief, details=err.details, resolution=err.resolution
+                )
+            )
+            return_code = 512
```
Comment (tigarmo) [inline craft_application/application.py:311]: where did these numbers come from?

Comment (lengau) [inline craft_application/application.py:311]:
```suggestion
        except craft_providers.ProviderError as err:
            self._emit_error(
                craft_cli.CraftError(
                    err.brief, details=err.details, resolution=err.resolution
                )
            )
```

I'm not certain how Python's implementation, Linux's implementation, and architecture endianness might interact with exit codes that don't fit in a byte. On my machine this ends up as 0 though, which isn't desired.

We'd have to do more work to sort out a decent [standard exit code](https://manpages.ubuntu.com/manpages/lunar/en/man3/sysexits.h.3head.html) from parts and provider errors, and that's a separate problem to tackle if we need to.

**Review by lengau — CHANGES_REQUESTED** [review]

File: craft_application/application.py (hunk 1 above)
Comment (sergiusens) [reply craft_application/application.py:311]: I think I copied them from an early version of craft application from @lengau ... alas I might be making that up!
I'll change it to 1

Comment (sergiusens) [reply craft_application/application.py:311]: catch

**Review by lengau — APPROVED** [review]
Thanks!

**Pull request merged**